File: Linked_List/merge_k_sorted_lists.py
"""
Merge k sorted linked lists and return it as one sorted list. Analyze and describe its complexity.

Example:

Input:
[
  1->4->5,
  1->3->4,
  2->6
]
Output: 1->1->2->3->4->4->5->6
"""

# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
import heapq
class Solution(object):
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        :rtype: ListNode
        """
        head = new = ListNode(0)
        heap = []
        
        for l in lists:
            if l:
                heapq.heappush(heap, (l.val, l))
        while heap:
            val, l = heapq.heappop(heap)
            new.next = ListNode(val)
            new = new.next
            if l.next:
                heapq.heappush(heap, (l.next.val, l.next))
        
        return head.next
        
        
        

File: java/Heap/merge_k_sorted_lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
        if (lists.length == 0) return null;
        PriorityQueue<ListNode> pq = new PriorityQueue<>(lists.length, new Comparator<ListNode>(){
            @Override
            public int compare(ListNode obj1, ListNode obj2){
                return obj1.val - obj2.val;
            }
        });
        
        for (ListNode temp: lists){
            if (temp!=null) pq.add(temp);
        }
        
        ListNode res = new ListNode(0);
        ListNode head = res;
        while(pq.size()!=0){
            ListNode temp = pq.poll();
            res.next = new ListNode(temp.val);
            res = res.next;
            if (temp.next!=null) pq.add(temp.next);
        }
        
        return head.next;
    }
}
